### PR TITLE
Add `Codata.Guarded.M` and update `Codata.Guarded.Stream`

### PR DESCRIPTION
**Pull request by XiaohuWang0921**
I have added `src/Codata/Guarded/M.agda` for M-types, and replaced `src/Codata/Guarded/Stream.agda` with my own version that I have been using, which contains some additional functions and alternative (and also possibly more elegant) implementations for old ones.

### CODE REVIEW

**Comment by gallais** [comment]
We'll need a merge & probably a fix of the proofs in `Codata.Guarded.Stream.Properties` before reviewing.

**Comment by gallais** [comment]
Seems like you copied the `Codata.Guarded.Stream.Properties` file & then modified it
rather than first merging `master` & then editing it. This has created even more merge
conflicts.

**Comment by MatthewDaggitt** [comment]
What's the status of this? Is it ready for review?

File: src/Codata/Guarded/Stream/Properties.agda [hunk 1]
```diff
@@ -190,6 +209,16 @@ lookup-tabulate : ∀ n (f : ℕ → A) → lookup n (tabulate f) ≡ f n
...
 lookup-tabulate (suc n) f = lookup-tabulate n (f ∘′ suc)
 
+lookup-transpose : ∀ n (ass : List (Stream A)) → lookup n (transpose ass) ≡ List.map (lookup n) ass
+lookup-transpose n [] = lookup-repeat n []
+lookup-transpose n (as ∷ ass) = begin
+  lookup n (transpose (as ∷ ass)) ≡⟨⟩
+  lookup n (zipWith _∷_ as (transpose ass)) ≡⟨ lookup-zipWith n _∷_ as (transpose ass) ⟩
+  lookup n as ∷ lookup n (transpose ass) ≡⟨ cong (lookup n as ∷_) (lookup-transpose n ass) ⟩
+  lookup n as ∷ List.map (lookup n) ass ≡⟨⟩
+  List.map (lookup n) (as ∷ ass) ∎
```
Comment (gallais) [inline src/Codata/Guarded/Stream/Properties.agda:219]:
```suggestion
lookup-transpose n (as ∷ ass) = begin
  lookup n (transpose (as ∷ ass))           ≡⟨⟩
  lookup n (zipWith _∷_ as (transpose ass)) ≡⟨ lookup-zipWith n _∷_ as (transpose ass) ⟩
  lookup n as ∷ lookup n (transpose ass)    ≡⟨ cong (lookup n as ∷_) (lookup-transpose n ass) ⟩
  lookup n as ∷ List.map (lookup n) ass     ≡⟨⟩
  List.map (lookup n) (as ∷ ass)            ∎
```

File: src/Codata/Guarded/Stream.agda [hunk 2]
```diff
@@ -34,118 +37,144 @@ record Stream (A : Set a) : Set a where
...
-tail (chunksOf n xs) = chunksOf n (drop n xs)
+infixr 5 _⁺++_
+_⁺++_ : List⁺ A → Stream A → Stream A
+(x ∷ xs) ⁺++ ys = x ∷ xs ++ ys
 
-interleave : Stream A → Stream A → Stream A
-head (interleave xs ys) = head xs
-tail (interleave xs ys) = interleave ys (tail xs)
+concat : Stream (List⁺ A) → Stream A
+concat {A = A} = λ s → [] ++concat s -- Here using λ so that _++concat_ doesn't take an extra argument
```
Comment (gallais) [inline src/Codata/Guarded/Stream.agda:117]:
```suggestion
concat {A = A} = [] ++concat_
```

File: src/Codata/Guarded/Stream.agda [hunk 3]
```diff
@@ -34,118 +37,144 @@ record Stream (A : Set a) : Set a where
...
+  module Concat where
+    _++concat_ : List A → Stream (List⁺ A) → Stream A
```
Comment (gallais) [inline src/Codata/Guarded/Stream.agda:119]: Why define an infix name if it's to use it suffix?

**Comment by MatthewDaggitt** [comment]
@XiaohuWang0921 do you agree with @gallais's changes?

**Comment by XiaohuWang0921** [comment]
@MatthewDaggitt Yes.

File: src/Codata/Guarded/Stream.agda (hunk 3 above)
Comment (MatthewDaggitt) [reply src/Codata/Guarded/Stream.agda:119]: How about you call it just `++-concat` and then call it using `++-concat []` on Line 117?

**Comment by MatthewDaggitt** [comment]
Great! Then could you implement them and add entries into `CHANGELOG.md` describing the changes? I'll then merge this in :+1:

**Pull request merged**